File: mobile/src/main/java/io/ashdavies/cinnamon/inject/ReleasableCallbacks.java
package io.ashdavies.cinnamon.inject;

import android.app.Application;
import android.content.ComponentCallbacks2;
import android.content.res.Configuration;
import dagger.releasablereferences.ReleasableReferenceManager;
import javax.inject.Inject;

public class ReleasableCallbacks implements ComponentCallbacks2 {

  private final ReleasableReferenceManager manager;

  @Inject
  ReleasableCallbacks(ReleasableReferenceManager manager) {
    this.manager = manager;
  }

  @Inject
  void register(Application application) {
    application.registerComponentCallbacks(this);
  }

  @Override
  public void onTrimMemory(int level) {
    if (level >= TRIM_MEMORY_RUNNING_MODERATE) {
      onLowMemory();
      return;
    }

    onHighMemory();
  }

  @Override
  public void onConfigurationChanged(Configuration config) {
    /* no op */
  }

  @Override
  public void onLowMemory() {
    manager.releaseStrongReferences();
  }

  private void onHighMemory() {
    manager.restoreStrongReferences();
  }
}

File: mobile/src/main/kotlin/io/ashdavies/cinnamon/inject/ReleasableCallbacks.kt
package io.ashdavies.cinnamon.inject

import android.app.Application
import android.content.ComponentCallbacks2
import android.content.res.Configuration
import dagger.releasablereferences.ReleasableReferenceManager
import javax.inject.Inject

class ReleasableCallbacks @Inject internal constructor(private val manager: ReleasableReferenceManager) : ComponentCallbacks2 {

  @Inject
  internal fun register(application: Application) {
    application.registerComponentCallbacks(this)
  }

  override fun onTrimMemory(level: Int) {
    if (level >= ComponentCallbacks2.TRIM_MEMORY_RUNNING_MODERATE) {
      onLowMemory()
      return
    }

    onHighMemory()
  }

  override fun onConfigurationChanged(config: Configuration) {
    /* no op */
  }

  override fun onLowMemory() {
    manager.releaseStrongReferences()
  }

  private fun onHighMemory() {
    manager.restoreStrongReferences()
  }
}
